src/swap/defi/thorchain/common.ts

```diff
@@ -570,39 +570,42 @@ export function makeThorchainBasedPlugin(
     }
 
     if (EVM_CURRENCY_CODES[fromMainnetCode]) {
+      if (router == null)
+        throw new Error(`Missing router address for ${fromMainnetCode}`)
+      if (thorAddress == null) {
+        throw new Error('Invalid vault address')
+      }
+
       memoType = 'hex'
+      let assetAddress = '0x0000000000000000000000000000000000000000' // mainnet
+      publicAddress = router
+
       if (fromTokenId != null) {
-        if (router == null)
-          throw new Error(`Missing router address for ${fromMainnetCode}`)
         if (sourceTokenContractAddress == null)
           throw new Error(
             `Missing sourceTokenContractAddress for ${fromMainnetCode}`
           )
-        // Need to use ethers.js to craft a proper tx that calls Thorchain contract, then extract the data payload
-        if (thorAddress == null) {
-          throw new Error('Invalid vault address')
-        }
-        memo = await getEvmTokenData({
-          assetAddress: sourceTokenContractAddress,
-          amountToSwapWei: Number(fromNativeAmount),
-          contractAddress: router,
-          vaultAddress: thorAddress,
-          memo
-        })
+        assetAddress = sourceTokenContractAddress
 
         // Token transactions send no ETH (or other EVM mainnet coin)
         ethNativeAmount = '0'
-        publicAddress = router
 
         // Check if token approval is required and return necessary data field
         approvalData = await getEvmApprovalData({
           contractAddress: router,
           assetAddress: sourceTokenContractAddress,
           nativeAmount: fromNativeAmount
         })
-      } else {
-        memo = Buffer.from(memo).toString('hex')
       }
+
+      // Need to use ethers.js to craft a proper tx that calls Thorchain contract, then extract the data payload
+      memo = await getEvmTokenData({
```

Comment:
Let's rename this function now that it's more generic, the name is misleading now because it's not specific to tokens